Implements an MBT prismatic joint

amcastro-tri opened the pull request:
Implements the internal workings to add translational dofs (`PrismaticMobilizer`) and, the higher level user facing API `PrismaticJoint`. 
A follow up PR introduces a cart pole model (with sdf parsing) and proper unit testing for that model.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8758)
<!-- Reviewable:end -->

amcastro-tri commented:
@avalenzu for feature review please.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBbk3BDDi1xWBevDE8i:bxshoqm)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
I believe you meant +@avalenzu for feature review. :-)

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBfGNwXEJ9i2qEm3zY_:b-dunjeq)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Checkpoint. Done with first two files. So far just lots of punctuation, word choice, and formatting nits.

---

Reviewed 2 of 8 files at r1.
Review status: 2 of 8 files reviewed at latest revision, all discussions resolved.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 18 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfGj_0705A9UE1_4yd:-LBfGj_0705A9UE1_4ye:bd1pdtn) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L18)):*
> ```objc
> namespace multibody {
> 
> /// This Mobilizer allows two frames to translate relatively to one another
> ```

BTW, this should be "relative to one another" - the whole thing is an adverbial phrase.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfIFWV-EIXYqUCLepV:-LBfIFWV-EIXYqUCLepW:byfq4in) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L19)):*
BTW

> in the direction of an axis that is constant when measured in either this mobilizer's inboard or outboard frames.

should be

> along an axis whose direction is constant when measured in either this mobilizer's inboard frame or its outboard frame.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 20 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfJ8ffD4VJxOHN15Eq:-LBfJ8ffD4VJxOHN15Er:bu3pmgn) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L20)):*
> ```objc
> /// This Mobilizer allows two frames to translate relatively to one another
> /// in the direction of an axis that is constant when measured in either this
> /// mobilizer's inboard or outboard frames. There is no relative translation
> ```

BTW s/translation/rotation/

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 22 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfJBQyDk8NeU5QuuCl:-LBfJBQyDk8NeU5QuuCm:bv7yguo) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L22)):*
> ```objc
> /// mobilizer's inboard or outboard frames. There is no relative translation
> /// between the inboard and outboard frames, just translation.
> /// To fully specify this mobilizer a user must provide the inboard frame F,
> ```

BTW, missing comma after "mobilizer." 

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfJQEP45NEW9nJn0Oy:-LBfJQEP45NEW9nJn0Oz:b-rgscdn) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L23)):*
> ```objc
> /// between the inboard and outboard frames, just translation.
> /// To fully specify this mobilizer a user must provide the inboard frame F,
> /// the outboard (or "mobilized") frame M and the axis `axis_F` (expressed in
> ```

BTW, "F", "M", and "`axis_f`" should have commas before and after them in the preceding part of this sentence (e.g. "the inboard frame, F,").

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 24 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfLVVq53r0Mb4cAHbd:-LBfLVVq53r0Mb4cAHbe:b-idmsfe) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L24)):*
> ```objc
> /// To fully specify this mobilizer a user must provide the inboard frame F,
> /// the outboard (or "mobilized") frame M and the axis `axis_F` (expressed in
> /// frame F) along which frame M translates with respect to F.
> ```

BTW, I would suggest using either "frame F/M" or "F/M" consistently. You're jumping back and forth.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 27 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfM8062aLmY7dwJtPg:-LBfM6xl-PDwINXcwd3d:b-7hm9mz) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L27)):*
> ```objc
> /// The single generalized coordinate q introduced by this mobilizer
> /// corresponds to the translation distance (in meters) of frame M with respect
> /// to frame F along the translation axis `axis_F`. When `q = 0`, frames F and M
> ```

BTW, ~the translation axis~

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfM_qT-I8lo_BVEolf:-LBfM_qT-I8lo_BVEolg:b-mhusnj) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L30)):*
BTW, I'd replace

> The translation distance is defined to be positive according in the direction of `axis_F` and it is defined negative in the direction opposite of `axis_F`.

with

> The translation distance is defined to be positive in the direction of `axis_F`.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 33 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfN6pV6N1u1p6MjDeT:-LBfN6pV6N1u1p6MjDeU:b-4kdiyu) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L33)):*
> ```objc
> /// Notice that the components of the translation axis as expressed in
> /// either frame F or M are constant. That is, `axis_F` and `axis_M` remain
> /// unchanged w.r.t. both frames by this mobilizer's motion.
> ```

BTW, these sentences ("Notice ... motion.") are redundant given your first sentence in the description.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 51 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfNa_B4wSJRTBJUaIc:-LBfNa_B4wSJRTBJUaId:blgh7ym) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L51)):*
BTW consider replacing

>  Constructor for a %PrismaticMobilizer between the inboard frame F `inboard_frame_F` and the outboard frame M `outboard_frame_F` granting a single translational degree of freedom along axis `axis_F`, expressed in the inboard frame F.

with

> Constructor for a %PrismaticMobilizer between the `inboard_frame_F` and `outboard_frame_M` granting a single translational degree of freedom along `axis_F`, expressed in the `inboard_frame_F`.

That way you can take advantage of having nicely readable variable names.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfO4rq8Oj6ZC6JgNqY:-LBfO4rq8Oj6ZC6JgNqZ:bagiz4c) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L52)):*
> ```objc
>   /// single translational degree of freedom along axis `axis_F`, expressed in
>   /// the inboard frame F.
>   /// @pre axis_F must be a non-zero vector with norm at least machine epsilon.
> ```

BTW s/axis_F/`axis_F`/

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 54 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfOG9117Y80REMFlX9:-LBfOG9117Y80REMFlXA:b-eb9cib) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L54)):*
BTW, do you need this, given the `@throws` statement on the following line?

> ```
>   /// This vector can have any length, only the direction is used. This method
>   /// throws an exception if the L2 norm of `axis` is less than machine epsilon.
> ```

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 73 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfTCzXBhk_Y2Y7KC92:-LBfTCzXBhk_Y2Y7KC93:ba4ioz3) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L73)):*
> ```objc
>   /// class documentation for sign convention details.
>   /// @throws std::logic_error if `context` is not a valid
>   /// MultibodyTreeContext.
> ```

BTW, Looks like this should fit on the previous line?

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfTcqR2abrik8I7NHa:-LBfTcqR2abrik8I7NHb:byqebn0) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L119)):*
> ```objc
>   /// along this mobilizer's axis (@see translation_axis().)
>   /// The generalized coordinate q for `this` mobilizer (the translation
>   /// distance) is stored in `context`.
> ```

BTW consider replacing "stored in" with "read from." Same for the next method.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 136 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfULcH6E5nx0k_KKqy:-LBfULcH6E5nx0k_KKqz:b-wkw63) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L136)):*
> ```objc
>   /// Computes the across-mobilizer acceleration `A_FM(q, v, v̇)` of the
>   /// outboard frame M in the inboard frame F.
>   /// By definition `A_FM = d_F(V_FM)/dt = H_FM(q) * v̇ + Ḣ_FM * v`.
> ```

Is the user supposed to know what `H_FM` is already?

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 147 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfUwp77AJBYlNBspFl:-LBfUwp77AJBYlNBspFm:b-94yirx) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L147)):*
> ```objc
> 
>   /// Projects the spatial force `F_Mo_F` on `this` mobilizer's outboard
>   /// frame M onto its translation axis (@see translation_axis().)
> ```

BTW, the `@see` command here causes strange formatting in the doxygen output.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBfVJbD4kSdm9PtFC91:b5uymfv)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review pass done. 

---

Reviewed 6 of 8 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 24 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfLVVq53r0Mb4cAHbd:-LBfjsXZ5Uq_OYbfJaKa:btmcx7x) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L24)):*
So this would either be

> along which M translates with respect to F.

or

> along which frame M translates with respect to frame F.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfTcqR2abrik8I7NHa:-LBfkBUHBi3AAnl7mbrA:b-e24xti) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L119)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW I know... I remember for instance Siyuan liking "store" better. Preferences I guess.
</blockquote></details>

BTW "stored in" makes it sound like this method stores it there.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 136 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfULcH6E5nx0k_KKqy:-LBfkQ527VBgFRKdflR5:bun8irb) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L136)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I don't know if I understand your question. However I do believe this doc could generally be moved to the `Mobilizer` class. I'll see to do that.
</blockquote></details>

I mean that there' s no explanation of what `H_FM` is here. It's not clear to me how the user is supposed to know what it is. 

---

*[multibody/multibody_tree/prismatic_mobilizer.cc, line 65 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfg7Kr-9AtU1emPObe:-LBfg7Ks6IgdP6jFP_ZF:bd3eve9) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.cc#L65)):*
> ```cpp
>     const MultibodyTreeContext<T>& context) const {
>   return Isometry3<T>(
>       Translation3<T>(get_translation(context) * translation_axis()));
> ```

BTW, You use `translation_axis()` here but `axis_F_` in the following two methods. The three usages should probably be consistent.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 15 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfdOEF8vzE2IPt6y6N:-LBfdOEF8vzE2IPt6y6O:b9nz96c) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L15)):*
> ```objc
> namespace drake {
> namespace multibody {
> ```

BTW, most of my comments from `prismatic_mobilizer.h` also apply to the documentation in this file.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 161 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfe20r7KEQEZ1JvQkD:-LBfe20r7KEQEZ1JvQkE:bpbrkfe) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L161)):*
> ```objc
>     // Right now we assume all the forces in joint_tau go into a single
>     // mobilizer.
>     DRAKE_DEMAND(joint_dof == 0);
> ```

Do we need to check this in Release builds given that this isn't a user-callable method?

---

*[multibody/multibody_tree/joints/test/prismatic_joint_test.cc, line 65 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfhzLJ19WLXhgSj9pz:-LBfhzLJ19WLXhgSj9q-:b-420qi4) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/test/prismatic_joint_test.cc#L65)):*
> ```cpp
> TEST_F(PrismaticJointTest, ContextDependentAccess) {
>   const double some_value = 1.5;
>   // Angle access:
> ```

BTW, these aren't angular quantities any more.

---

*[multibody/multibody_tree/test/prismatic_mobilizer_test.cc, line 75 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfiLxV0Vi745ZkTEsI:-LBfiLxW9R34N1Eh5sCt:b-r1lx9c) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/test/prismatic_mobilizer_test.cc#L75)):*
> ```cpp
> }
> 
> // Verifies method to mute and access the context.
> ```

BTW s/mute/mutate

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBfjlvF7RkiFGWclzDt:b-c312rk)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a bunch of pedantic BTWs

---

Review status: all files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBfns1VBGecspEbRJGO:b-q0y5vn)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Tank u! +@sherm1 for platform please.

---

Review status: all files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBfoNHI2k7P8duC-649:boa35ao)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint. Please generate the doxygen and inspect it to make sure everything formats as you intended.

---

Reviewed 5 of 8 files at r1.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfJQEP45NEW9nJn0Oy:-LBl_A6yBkgrqejUrA0i:b-i90h0y) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L23)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, "F", "M", and "`axis_f`" should have commas before and after them in the preceding part of this sentence (e.g. "the inboard frame, F,").
</blockquote></details>

FYI the extra commas would be inconsistent with our other documentation. I'm not sure they would add anything here (though they make grammatical sense). My impression is that offsetting symbols by commas is uncommon in mathematical text?

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 26 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBla9fKCWn3OcYHpLUZ:-LBla9fKCWn3OcYHpLU_:b-19mwku) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L26)):*
> ```objc
> /// frame F) along which frame M translates with respect to F.
> /// The single generalized coordinate q introduced by this mobilizer
> /// corresponds to the translation distance (in meters) of frame M with respect
> ```

BTW consider "of frame M" -> "of the origin Mo of frame M"

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlarm97VAIccQruoY_:-LBlarm97VAIccQruoYa:bk0aoai) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L49)):*
> ```objc
> 
>   /// Constructor for a %PrismaticMobilizer between the inboard frame F
>   /// `inboard_frame_F` and the outboard frame M `outboard_frame_F` granting a
> ```

BTW typo: outboard_frame_F -> outboard_frame_M

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 51 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfNa_B4wSJRTBJUaIc:-LBlb4cX4oaqJG6Ed--p:bb6mt9b) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L51)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW consider replacing

>  Constructor for a %PrismaticMobilizer between the inboard frame F `inboard_frame_F` and the outboard frame M `outboard_frame_F` granting a single translational degree of freedom along axis `axis_F`, expressed in the inboard frame F.

with

> Constructor for a %PrismaticMobilizer between the `inboard_frame_F` and `outboard_frame_M` granting a single translational degree of freedom along `axis_F`, expressed in the `inboard_frame_F`.

That way you can take advantage of having nicely readable variable names.

</blockquote></details>

:+1:

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlbgWC9KiXzdhNVaZi:-LBlbgWC9KiXzdhNVaZj:b-s5j8p3) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L52)):*
> ```objc
>   /// single translational degree of freedom along axis `axis_F`, expressed in
>   /// the inboard frame F.
>   /// @pre axis_F must be a non-zero vector with norm at least machine epsilon.
> ```

BTW that is a very generous rule. It seems likely to me that such a small norm really is just noise and the user accidently used a meaningless axis vector here. Consider being more strict and requiring a length that is clearly not noise, like √ε. In most cases I think the user will be glad to get an error message for an "axis" that is basically noise. If they _really_ mean to use such a short vector they can pre-normalize it for you.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 55 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlbIUj9sv51DOdmZGU:-LBlbIUkDGY4DrOJyAp3:bzehewo) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L55)):*
> ```objc
>   /// This vector can have any length, only the direction is used. This method
>   /// throws an exception if the L2 norm of `axis` is less than machine epsilon.
>   /// @throws std::exception if the L2 norm of `axis` is less than machine
> ```

BTW axis -> axis_F

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 99 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlcubOCkbniDNlIjRZ:-LBlcubOCkbniDNlIjR_:b86jvql) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L99)):*
> ```objc
>   const T& get_translation_rate(const systems::Context<T> &context) const;
> 
>   /// Sets the rate of change, in meters per second, of this `this` mobilizer's
> ```

BTW this this -> this

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 100 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBld-SlAklEQgC1CPZP:-LBld-SlAklEQgC1CPZQ:b-4zh3ss) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L100)):*
> ```objc
> 
>   /// Sets the rate of change, in meters per second, of this `this` mobilizer's
>   /// translation to `translation_dot`. The new rate of change `theta_dot` gets
> ```

BTW copy pasta (no theta here)

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 113 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBldsG31BcrWqGcvYcb:-LBldsG31BcrWqGcvYcc:b8pu0ob) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L113)):*
> ```objc
>   /// Sets `state` to store a zero translation and translational rate.
>   void set_zero_state(const systems::Context<T>& context,
>                       systems::State<T>* state) const override;
> ```

BTW META: consider using `final` rather than `override` when overriding methods in a final class. That way it is clear just looking at the declaration here that the method cannot be further extended (otherwise it requires looking far away).

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 124 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlf6n15jH3JeGCE3yk:-LBlf6n15jH3JeGCE3yl:b-j6ujc4) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L124)):*
> ```objc
> 
>   /// Computes the across-mobilizer velocity `V_FM(q, v)` of the outboard frame
>   /// M measured and expressed in frame F as a function of the translation and
> ```

BTW translation -> translation taken from `context` (to match the acceleration doc below)

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 132 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBleGWf-eWd_zOU3xob:-LBleGWf-eWd_zOU3xoc:b48bl0r) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L132)):*
> ```objc
>   SpatialVelocity<T> CalcAcrossMobilizerSpatialVelocity(
>       const MultibodyTreeContext<T>& context,
>       const Eigen::Ref<const VectorX<T>>& v) const override;
> ```

FYI (not for this PR): it seems odd to me that there is no method that returns V_FM given just a Context. That would generally be cached, while this method must recalculate V_FM even if v actually comes from `context`. This method is more flexible though since it can use any v, so no need for another method if caching is not an issue here.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 147 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfUwp77AJBYlNBspFl:-LBlfjaiBIxtAj6qJ-ag:bsibq8s) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L147)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, the `@see` command here causes strange formatting in the doxygen output.
</blockquote></details>

FYI `@see` should be on its own line, usually at the end of a doxygen block. Please be sure to generate the doxygen and inspect it closely to make sure it is formatted as you intended it. 

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 164 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlg5hIANFu6Ulceced:-LBlg5hIANFu6Ulcecee:bqsvxll) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L164)):*
> ```objc
>       EigenPtr<VectorX<T>> qdot) const override;
> 
>   void MapQDotToVelocity(
> ```

BTW this and the previous method should have a short doxygen doc explaining that for this mobilizer qdot=v.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 192 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlgbAVBtCllKnc7ipa:-LBlgbAVBtCllKnc7ipb:b-23q11q) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L192)):*
> ```objc
>       const MultibodyTree<ToScalar>& tree_clone) const;
> 
>   // Default axis expressed in the inboard frame F.
> ```

BTW mention that it is a unit vector.

---

*[multibody/multibody_tree/prismatic_mobilizer.cc, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlkDAw1F8QsN6tlun8:-LBlkDAw1F8QsN6tlun9:bbl5w7p) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.cc#L19)):*
> ```cpp
>   auto q = this->get_positions(mbt_context);
>   DRAKE_ASSERT(q.size() == kNq);
>   return q.coeffRef(0);
> ```

BTW can you write`q[0]` as you did below?

---

*[multibody/multibody_tree/prismatic_mobilizer.cc, line 40 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlkNi_81FicYwRADan:-LBlkNi_81FicYwRADao:b1gr29d) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.cc#L40)):*
> ```cpp
>   const auto& v = this->get_velocities(mbt_context);
>   DRAKE_ASSERT(v.size() == kNv);
>   return v.coeffRef(0);
> ```

BTW v[0] ?

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 57 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlhEil0kLMgjYzfByI:-LBlhEil0kLMgjYzfByJ:b754brv) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L57)):*
> ```objc
>   ///   This vector can have any length, only the direction is used.
>   /// @throws std::exception if the L2 norm of `axis` is less than machine
>   /// epsilon.
> ```

BTW see my comment for the mobilizer.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 84 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlhdPNCMKDMXrQEIzW:-LBlhdPNCMKDMXrQEIzX:bgs24p0) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L84)):*
> ```objc
>   /// @param[in] context
>   ///   The context of the MultibodyTree this joint belongs to.
>   /// @returns The transaltion coordinate of `this` joint stored in the
> ```

BTW typo

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 113 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlhtL03RHRY5pNQ4xA:-LBlhtL03RHRY5pNQ4xB:b86jvql) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L113)):*
> ```objc
>   }
> 
>   /// Sets the rate of change, in meters per second, of this `this` joint's
> ```

BTW this this -> this

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 128 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBli6fA-38WfzuSPa3k:-LBli6fA-38WfzuSPa3l:bapbl37) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L128)):*
> ```objc
>   }
> 
>   /// @}
> ```

FYI `//@}` works if you want something shorter.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 158 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBljEh9AaAKntEcAgoA:-LBljEh9AaAKntEcAgoB:b-lzu2jh) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L158)):*
> ```objc
>       int joint_dof,
>       const T& joint_tau,
>       MultibodyForces<T>* forces) const override {
> ```

BTW META: consider `final` for overrides here since the class is final.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 165 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBliu3qBRPDndu-otZ9:-LBliu3rEw1CA1ufDBnF:bf0r9xh) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L165)):*
> ```objc
>         get_mobilizer()->get_mutable_generalized_forces_from_array(
>             &forces->mutable_generalized_forces());
>     tau_mob(joint_dof) += joint_tau;
> ```

BTW I suspect the above is generating a lot of code for simply adding in an actuation force to an element of an existing array. Might be worth a TODO here to consider a cheaper method for adding in a single scalar force, which is very common.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 215 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBljh8L2fJue53FIs-V:-LBljh8L2fJue53FIs-W:b-1zig1y) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L215)):*
> ```objc
>       const MultibodyTree<ToScalar>& tree_clone) const;
> 
>   // This is the joint's axis expressed in either M or F since axis_M = axis_F.
> ```

BTW mention that it's a unit vector

---

*[multibody/multibody_tree/joints/prismatic_joint.cc, line 27 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBllWqw9bAVFHMkzQBu:-LBllWqw9bAVFHMkzQBv:b-qkdden) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.cc#L27)):*
> ```cpp
>       this->translation_axis());
> 
>   return std::move(joint_clone);
> ```

BTW move not needed here (and better not to use it).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBlfNqs5dpUAdozheG_:bw41e94)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Beautiful! Platform pass done with just minor comments, PTAL. 
Also, if you haven't done so already please run kcov to make sure that all the new code has unit tests.

---

Reviewed 3 of 8 files at r1.
Review status: all files reviewed at latest revision, 38 unresolved discussions.

---

*[multibody/multibody_tree/joints/test/prismatic_joint_test.cc, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlodnN6GujCPZ8GyJA:-LBlodnN6GujCPZ8GyJB:b3txh6k) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/test/prismatic_joint_test.cc#L3)):*
> ```cpp
> // clang-format: off
> #include "drake/multibody/multibody_tree/multibody_tree.h"
> // clang-format: on
> ```

BTW why not put prismatic_joint.h here and multibody_tree.h below?

---

*[multibody/multibody_tree/joints/test/prismatic_joint_test.cc, line 69 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlovLnD1wdMiU0JTGK:-LBlovLnD1wdMiU0JTGL:bmnecwj) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/test/prismatic_joint_test.cc#L69)):*
> ```cpp
>   EXPECT_EQ(joint1_->get_translation(*context_), some_value);
> 
>   // Angular rate access:
> ```

BTW angle

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 202 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlmqpE6xvKmAPXfoMc:-LBlmqpE6xvKmAPXfoMd:b-x8kgcy) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L202)):*
> ```cpp
>     case sdf::JointType::PRISMATIC: {
>       Vector3d axis_J = ExtractJointAxis(joint_spec);
>       plant->AddJoint<PrismaticJoint>(
> ```

BTW is there a unit test for the parser that shows it parses a prismatic joint successfully? Please add if not.

---

*[multibody/multibody_tree/test/prismatic_mobilizer_test.cc, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlmIhxALqwnFKDmDVw:-LBlmIhxALqwnFKDmDVx:b-y22esr) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/test/prismatic_mobilizer_test.cc#L3)):*
> ```cpp
> // clang-format: off
> #include "drake/multibody/multibody_tree/multibody_tree.h"
> // clang-format: on
> ```

BTW why not put prismatic_mobilizer.h here and multibody_tree.h down below? Then you wouldn't need the clang-format guards and it would be clearer what's being tested.

---

*[multibody/multibody_tree/test/prismatic_mobilizer_test.cc, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlnNi55OHEEWcAeOX8:-LBlnNi55OHEEWcAeOX9:b964hrq) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/test/prismatic_mobilizer_test.cc#L53)):*
> ```cpp
>     // Performance critical queries take a MultibodyTreeContext to avoid dynamic
>     // casting.
>     mbt_context_ = dynamic_cast<MultibodyTreeContext<double>*>(context_.get());
> ```

BTW consider `EXPECT_NE(mbt_context_, nullptr)`

---

*[multibody/multibody_tree/test/prismatic_mobilizer_test.cc, line 78 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlnjB-BOdX8tOCF98X:-LBlnjB-BOdX8tOCF98Y:b-f0l0ua) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/test/prismatic_mobilizer_test.cc#L78)):*
> ```cpp
> TEST_F(PrismaticMobilizerTest, StateAccess) {
>   const double some_value1 = 1.5;
>   const double some_value2 = M_PI;
> ```

BTW this choice seems slightly misleading since there is no angle here. Consider something more translation-y to avoid confusion ? sqrt(2)? (also below)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LBlpJST6AwUhB1BqkO9:b-50ar88)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfJQEP45NEW9nJn0Oy:-LBvce9f30vCyvnlT-bb:b-3a5igm) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L23)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

OK I'll defer to you on this, @sherm1 .
</blockquote></details>

@avalenzu explained this "grammar" rule at lunch time, TIL. However, I do like the version without the commands for these comments.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 24 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfLVVq53r0Mb4cAHbd:-LBvd3DcAlowFRXawu2g:b-s622lo) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L24)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

So this would either be

> along which M translates with respect to F.

or

> along which frame M translates with respect to frame F.
</blockquote></details>

I see, thanks

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfTcqR2abrik8I7NHa:-LBvh8wB52wt3n8K0Qsb:b2gg17b) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L119)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW "stored in" makes it sound like this method stores it there.
</blockquote></details>

Ah! yes. Thanks.

---

*[multibody/multibody_tree/prismatic_mobilizer.h, line 136 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfULcH6E5nx0k_KKqy:-LBviRsgCDT27fYk9A7G:b-sjjdxs) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.h#L136)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I mean that there' s no explanation of what `H_FM` is here. It's not clear to me how the user is supposed to know what it is. 
</blockquote></details>

Right. I decided to remove H_FM from this discussion since there are details in the `Mobilizer` class.

---

*[multibody/multibody_tree/prismatic_mobilizer.cc, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlkDAw1F8QsN6tlun8:-LBw3uRx2DCtLkSOGL-n:b-recmrw) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.cc#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW can you write`q[0]` as you did below?
</blockquote></details>

BTW. There is a reason for this, and is that returning `q[0]` causes the error:
```
returning reference to local temporary object [-Werror,-Wreturn-stack-address]`
```

I don't understand why Eigen decides to return a copy with the `const operator[]() const`, but it does. It does however what you expect with `coefRef()`. I never dug dipper on why this would be.

---

*[multibody/multibody_tree/prismatic_mobilizer.cc, line 40 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlkNi_81FicYwRADan:-LBw4cG-8RdNJ16U9B72:b-z3p5ge) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/prismatic_mobilizer.cc#L40)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW v[0] ?
</blockquote></details>

ditto above

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 161 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBfe20r7KEQEZ1JvQkD:-LBw7tlU2b5Xe4-ki3jK:b-a0ac4s) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/joints/prismatic_joint.h#L161)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Do we need to check this in Release builds given that this isn't a user-callable method?
</blockquote></details>

True that. Removed.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 202 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LBlmqpE6xvKmAPXfoMc:-LBw9Uft7ljpbUkQXoPu:b-xsvq84) ([raw file](https://github.com/robotlocomotion/drake/blob/b91c6595a6c2690c973d3ddb12134d9e1070da73/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L202)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is there a unit test for the parser that shows it parses a prismatic joint successfully? Please add if not.
</blockquote></details>

BTW No reason to introduce here since I have planed to push a cart pole model to test this anyway. So I reverted this change for this PR.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Reviewed 6 of 7 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (added one minor FYI).

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/multibody_tree/joints/prismatic_joint.h, line 145 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8758#-LC--jljDM1UAb3no_Tt:-LC--jljDM1UAb3no_Tu:bnq0diw) ([raw file](https://github.com/robotlocomotion/drake/blob/5061f79f8096a760c86d83eda505e06892120a06/multibody/multibody_tree/joints/prismatic_joint.h#L145)):*
> ```objc
> 
>  protected:
>   /// Joint<T> final called through public NVI, Joint::AddInForce().
> ```

FYI this should be called a "virtual" (final & override are just the keywords).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8758#-:-LC-0DP-1QxUsHuj0ljZ:byw42xm)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks @sherm1, I already added this correction to my next PR in the queue (a cart pole demo using this joint).
